src/lib/key_material.cpp

```diff
@@ -619,9 +734,10 @@ DSAKeyMaterial::write_secret(pgp_packet_body_t &pkt) const
 }
 
 bool
-DSAKeyMaterial::generate(const rnp_keygen_crypto_params_t &params)
+DSAKeyMaterial::generate(rnp::SecurityContext &ctx, const KeyParams &params)
 {
-    if (dsa_generate(&params.ctx->rng, &key_, params.dsa.p_bitlen, params.dsa.q_bitlen)) {
+    auto &dsa = dynamic_cast<const DSAKeyParams &>(params);
```

Comment:
The idea was to avoid theoretical possibility of getting wrong object here by throwing exception. Not sure how that could happen, just to be cautious.